ANPR/src/anpr/license_plate_recognition.py

```diff
@@ -0,0 +1,283 @@
+
+    def __init__(self, text: str, location: ResultLocation):
+        self.text = text
+        self.location = location
+
+    def match_format(self, formats: list):
+        """
+        This function matches the detected text to the given regex patterns. It also creates new LicensePlateResult's
+        based on the matching patterns.
+        @param formats: list for regex patterns
+        @return: list of LicensePlateResult's with matched strings
+        """
+
+        # Make list for new results
+        matches = []
+        # Go over each regex format and find all matching strings
+        for f in formats:
+            matches += re.findall(f, self.text)
+
+        return [LicensePlateResult(text, self.location) for text in matches]
+
+
+class ANPR:
+    """
+    This class can perform Automatic Number Plate Recognition (ANPR) on given images.
+    """
+
+    def __init__(self, selection_ocr: OCR, result_ocr: OCR, formats: list[str] = None, minAR=3,
+                 maxAR=7, verbosity: int = 4, ):
+        # Store the minimum and maximum rectangular aspect ratio
+        self.minAR = minAR
+        self.maxAR = maxAR
+
+        # Variable to store how much debug output the class should give
+        #        Levels of verbosity:
+        #   * -1: nothing
+        #   *  0: prints detected license plates
+        #   *  1: show result image
+        #   *  2: show all detected license plate candidates
+        #   *  3: show result of black/white transformation to find license plate shapes
+        #   *  4: show all steps
+
+        self.verbosity = verbosity
+
+        # Text recognition classes to use for selecting license plates out of multiple candidates and for giving the
+        # final result
+        self.selection_ocr = selection_ocr
+        self.result_ocr = result_ocr
+
+        # Transform custom format to regex ex. 'N-LLL-NNN' -> '[0-9]-[A-Z][A-Z][A-Z]-[0-9][0-9][0-9]'
+        if formats is not None:
+            formats = [f.replace('N', '[0-9]') for f in formats]
+            formats = [f.replace('L', '[A-Z]') for f in formats]
+            self.debug_print(formats, verbosity=0)
+        self.formats = formats
+
+    def debug_imshow(self, title, image, verbosity: int, waitKey=False):
+        """
+        Method to show debug or result images.
+
+
+        @param title: Title of image window
+        @param image: Image to show
+        @param verbosity: level of verbosity required to show the image
+        @param waitKey: boolean indicating whether the program should sleep until the user presses a key.
+        """
+
+        # Check if the verbosity of the image is below the verbosity specified by the user. A low verbosity means the
+        # image is more important
+        if verbosity <= self.verbosity:
+            cv2.imshow(title, image)
+            # Check to see if we should wait for a keypress
+            if waitKey:
+                cv2.waitKey(0)
+
+    def debug_print(self, text, verbosity):
```

Comment:
Nit: maybe mark the output type of the function: `def debug_print(self, text:  str, verbosity: int) -> None`.